File: FlowCrypt/src/main/java/com/flowcrypt/email/util/graphics/glide/transformations/CircleTransformation.java
/*
 * © 2016-2018 FlowCrypt Limited. Limitations apply. Contact human@flowcrypt.com
 * Contributors: DenBond7
 */

package com.flowcrypt.email.util.graphics.glide.transformations;

import android.graphics.Bitmap;
import android.graphics.BitmapShader;
import android.graphics.Canvas;
import android.graphics.Paint;

import com.bumptech.glide.load.engine.bitmap_recycle.BitmapPool;
import com.bumptech.glide.load.resource.bitmap.BitmapTransformation;

import java.security.MessageDigest;

import androidx.annotation.NonNull;

/**
 * The circle transformation for the Glide.
 * For details see https://stackoverflow
 * .com/questions/25278821/how-to-round-an-image-with-glide-library/25806229#25806229
 *
 * @author Denis Bondarenko
 * Date: 17.07.2017
 * Time: 15:59
 * E-mail: DenBond7@gmail.com
 */

public class CircleTransformation extends BitmapTransformation {

  @Override
  public void updateDiskCacheKey(MessageDigest messageDigest) {

  }

  @Override
  protected Bitmap transform(@NonNull BitmapPool bitmapPool, @NonNull Bitmap bitmap,
                             int i, int i1) {
    return circleCrop(bitmapPool, bitmap);
  }

  public String getId() {
    return getClass().getName();
  }

  private static Bitmap circleCrop(BitmapPool pool, Bitmap source) {
    if (source == null) return null;

    int size = Math.min(source.getWidth(), source.getHeight());
    int x = (source.getWidth() - size) / 2;
    int y = (source.getHeight() - size) / 2;

    Bitmap squared = Bitmap.createBitmap(source, x, y, size, size);
    Bitmap result = pool.get(size, size, Bitmap.Config.ARGB_8888);

    Canvas canvas = new Canvas(result);
    Paint paint = new Paint();
    paint.setShader(new BitmapShader(squared, BitmapShader.TileMode.CLAMP, BitmapShader
        .TileMode.CLAMP));
    paint.setAntiAlias(true);
    float r = size / 2f;
    canvas.drawCircle(r, r, r, paint);
    return result;
  }
}

File: FlowCrypt/src/main/java/com/flowcrypt/email/util/graphics/glide/transformations/CircleTransformation.kt
/*
 * © 2016-present FlowCrypt a.s. Limitations apply. Contact human@flowcrypt.com
 * Contributors: DenBond7
 */

package com.flowcrypt.email.util.graphics.glide.transformations

import android.graphics.Bitmap
import android.graphics.BitmapShader
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Shader

import com.bumptech.glide.load.engine.bitmap_recycle.BitmapPool
import com.bumptech.glide.load.resource.bitmap.BitmapTransformation

import java.security.MessageDigest

/**
 * The circle transformation for the Glide.
 * For details see https://stackoverflow
 * .com/questions/25278821/how-to-round-an-image-with-glide-library/25806229#25806229
 *
 * @author Denys Bondarenko
 */
class CircleTransformation : BitmapTransformation() {

  val id: String
    get() = javaClass.name

  override fun updateDiskCacheKey(messageDigest: MessageDigest) {

  }

  override fun transform(bitmapPool: BitmapPool, bitmap: Bitmap, i: Int, i1: Int): Bitmap? {
    return circleCrop(bitmapPool, bitmap)
  }

  private fun circleCrop(pool: BitmapPool, source: Bitmap?): Bitmap? {
    if (source == null) {
      return null
    }

    val size = Math.min(source.width, source.height)
    val x = (source.width - size) / 2
    val y = (source.height - size) / 2

    val squared = Bitmap.createBitmap(source, x, y, size, size)
    val result = pool.get(size, size, Bitmap.Config.ARGB_8888)

    val canvas = Canvas(result)
    val paint = Paint()
    paint.shader = BitmapShader(squared, Shader.TileMode.CLAMP, Shader.TileMode.CLAMP)
    paint.isAntiAlias = true
    val r = size / 2f
    canvas.drawCircle(r, r, r, paint)
    return result
  }
}
